### ISSUE TITLE
Add org.jooq.Period for temporal validity

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#9741]
When supporting temporal validity in jOOQ (#4704), we need a new `org.jooq.Period` meta model, which describes periods that can be searched using new API on `org.jooq.Table` and elsewhere.

There are two types of periods:

- System versioned
- Application versioned

The period types can be combined for bi-temporal versioning.

In standard SQL, application versioned periods can be named, whereas the system period is called `SYSTEM_TIME`.

This issue includes:

- [x] The period definition
- [x] DSL API to create period instances
- [x] API to reference periods from tables (and thus from queries)
- [x] Translations to different dialects, currently including:
  - [x] DB2
  - [x] MariaDB
  - [x] Oracle
  - [x] SQL Server
- [x] Parser support

**Comment by lukaseder** [comment]
A draft API combining meta model with DSL (draft will be updated):

```java
public interface Period<T> {

    Name getName();

    Field<T> getStart();
    Field<T> getEnd();

    // SQL standard clauses
    PeriodSpecification asOf(T value);
    PeriodSpecification asOf(Field<T> field);

    PeriodSpecification between(T value1, T value2);
    PeriodSpecification between(Field<T> field1, Field<T> field2);
    PeriodSpecificationBetweenAndStep<T> between(T value1);
    PeriodSpecificationBetweenAndStep<T> between(Field<T> field1);

    PeriodSpecification from(T value1, T value2);
    PeriodSpecification from(Field<T> field1, Field<T> field2);
    PeriodSpecificationFromToStep<T> from(T value1);
    PeriodSpecificationFromToStep<T> from(Field<T> field1);

    // Oracle clause
    PeriodSpecificationBetweenAndStep<T> betweenMinvalue();

    // SQL Server clause
    PeriodSpecification containedIn(T value1, T value2);
    PeriodSpecification containedIn(Field<T> field1, Field<T> field2);

    // MariaDB clause
    PeriodSpecification all();

    interface PeriodSpecification {}
    interface PeriodSpecificationBetweenAndStep<T> {

        // SQL standard clauses
        PeriodSpecification and(T value2);
        PeriodSpecification and(Field<T> field2);

        // Oracle clause
        PeriodSpecification andMaxvalue();
    }

    interface PeriodSpecificationFromToStep<T> {

        // SQL standard clauses
        PeriodSpecification to(T value2);
        PeriodSpecification to(Field<T> field2);
    }
}
``` 

And then, in `Table`:

```java
    // SQL standard
    Table<R> for_(Period.PeriodSpecification periodSpecification);
    Table<R> forPortionOf(Period.PeriodSpecification periodSpecification);

    // Oracle specific
    <T> Table<R> asOfPeriodFor(Period<T> period, T value);
    <T> Table<R> asOfPeriodFor(Period<T> period, Field<T> field);
    Table<R> versionsPeriodFor(Period.PeriodSpecification periodSpecification);
```

**Comment by lukaseder** [comment]
This is now implemented for jOOQ 3.13. Based on this API, we can implement code generation and emulations, and much more in the future.

**Comment by lukaseder** [comment]
HSQLDB supports temporal tables as well: #9967
